Index: libapache-mod-fastcgi/fcgi_protocol.h
===================================================================
--- libapache-mod-fastcgi.orig/fcgi_protocol.h	2011-09-01 16:46:57.659269391 +0200
+++ libapache-mod-fastcgi/fcgi_protocol.h	2011-09-01 16:48:37.052794371 +0200
@@ -52,7 +52,8 @@
 #define FCGI_DATA                8
 #define FCGI_GET_VALUES          9
 #define FCGI_GET_VALUES_RESULT  10
-#define FCGI_UNKNOWN_TYPE       11
+#define FCGI_BYTE_ACC           15
+#define FCGI_UNKNOWN_TYPE       16
 #define FCGI_MAXTYPE (FCGI_UNKNOWN_TYPE)
 
 /*
Index: libapache-mod-fastcgi/fcgi_protocol.c
===================================================================
--- libapache-mod-fastcgi.orig/fcgi_protocol.c	2011-09-01 16:48:43.719810672 +0200
+++ libapache-mod-fastcgi/fcgi_protocol.c	2011-09-01 16:49:21.150740683 +0200
@@ -478,6 +478,8 @@
                     fr->readingEndRequestBody = FALSE;
                 }
                 break;
+	    case FCGI_BYTE_ACC:
+		/* XXX coming soon */
             case FCGI_GET_VALUES_RESULT:
                 /* XXX coming soon */
             case FCGI_UNKNOWN_TYPE:
