Index: libapache-mod-fastcgi/mod_fastcgi.c
===================================================================
--- libapache-mod-fastcgi.orig/mod_fastcgi.c	2011-08-30 16:24:52.243055659 +0200
+++ libapache-mod-fastcgi/mod_fastcgi.c	2011-08-30 16:26:22.672476702 +0200
@@ -719,7 +719,7 @@
             int statusValue = strtol(value, NULL, 10);
 
             if (hasStatus) {
-                goto DuplicateNotAllowed;
+		continue;
             }
             if (statusValue < 0) {
                 fr->parseHeader = SCAN_CGI_BAD_HEADER;
@@ -734,7 +734,7 @@
         if (fr->role == FCGI_RESPONDER) {
             if (strcasecmp(name, "Content-type") == 0) {
                 if (hasContentType) {
-                    goto DuplicateNotAllowed;
+		    continue;
                 }
                 hasContentType = TRUE;
 #ifdef APACHE2                
@@ -747,7 +747,7 @@
 
             if (strcasecmp(name, "Location") == 0) {
                 if (hasLocation) {
-                    goto DuplicateNotAllowed;
+		    continue;
                 }
                 hasLocation = TRUE;
                 ap_table_set(r->headers_out, "Location", value);
